### PR TITLE
Output GTFS link mappings CSV file during GTFS link mapping process

### PR DESCRIPTION
**Pull request by mark-idleman**

### CODE REVIEW

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java [hunk 1]
```diff
@@ -202,4 +214,69 @@ public void setGtfsLinkMappings() {
...
     }
+
+    // returns all CSV rows (as a list of Strings) derived from a single GTFS feed's data
+    private List<String> getGtfsLinkCsvRowsForFeed(SetMultimap<String, String> routeIdToTripsInRoute,
+                                                   Map<String, List<Pair<Stop, Stop>>> tripIdToStopPairsInTrip,
+                                                   HTreeMap<String, String> gtfsLinkMappings) {
+        List<String> rowsForFeed = Lists.newArrayList();
+        for (String routeId : routeIdToTripsInRoute.keySet()) {
+            for (String tripIdInRoute : routeIdToTripsInRoute.get(routeId)) {
+                for (Pair<Stop, Stop> stopStopPair : tripIdToStopPairsInTrip.get(tripIdInRoute)) {
```
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:225]: todo: sanity check for stop/stop pairs that are the same

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java [hunk 2]
```diff
@@ -62,6 +60,10 @@ public void setGtfsLinkMappings() {
                 .valueSerializer(Serializer.STRING)
                 .make();
 
+        // Output file location for CSV containing all GTFS link mappings
+        File gtfsLinksCsvOutput = new File(graphHopper.getGraphHopperLocation() + "/gtfs_link_mappings.csv");
```
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:64]: todo: change to `gtfs_link_mapping.csv`

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java [hunk 3]
```diff
@@ -202,4 +214,69 @@ public void setGtfsLinkMappings() {
...
+            for (String tripIdInRoute : routeIdToTripsInRoute.get(routeId)) {
+                for (Pair<Stop, Stop> stopStopPair : tripIdToStopPairsInTrip.get(tripIdInRoute)) {
+                    String stopPairString = stopStopPair.getLeft().stop_id + "," + stopStopPair.getRight().stop_id;
+
+                    // Skip stop-stop pairs where we couldn't find a valid route
+                    if (!gtfsLinkMappings.containsKey(stopPairString)) {
+                        continue;
+                    }
+                    List<String> stableEdgeIds = Lists.newArrayList(gtfsLinkMappings.get(stopPairString).split(","));
+                    String stableEdgeIdString = String.format("\"[%s]\"", stableEdgeIds.stream().map(id -> "'" + id + "'").collect(Collectors.joining(",")));
```
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:233]: todo: output `""` for this if we don't have any stable edge IDs. So, the CSV column will look like `,,` basically

**Review by kaelgreco — DISMISSED** [review]
lgtm 👍 

what dm is this part of? should probably add a source to the configs so it's easy to find

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java [hunk 4]
```diff
@@ -202,4 +214,74 @@ public void setGtfsLinkMappings() {
...
+                    List<String> stableEdgeIds = Lists.newArrayList(gtfsLinkMappings.get(stopPairString).split(","));
+                    String stableEdgeIdString = stableEdgeIds.size() == 0 ? ""
+                            : String.format("\"[%s]\"", stableEdgeIds.stream().map(id -> "'" + id + "'").collect(Collectors.joining(",")));
+
+                    Stop stop = stopStopPair.getLeft();
+                    Stop nextStop = stopStopPair.getRight();
+
+                    // todo: why does this exist? Just copied over logic from old link mapper for now
+                    // format: "{feed_id}:{route_id}/{feed_id}:{stop_id}/{feed_id}:{next_stop_id}"
+                    String transitEdgeString = stop.feed_id + ":" + routeId + "/" + stop.feed_id + ":"
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:245]: Ahh shoot, forgot to remove this todo - this makes sense! The picture got clearer after working on the stop_id fix for Brett as well. I'll remove the todo and keep this as-is, I spot checked a CSV generated by this code and it looks like it's using the "correct" feed IDs here, so we should be good to go

**Comment by mark-idleman** [comment]
@kaelgreco it's part of the `GraphHopperBuilder` dm (ie, the one that runs the link mapper) - I already coordinated w/ Aditya and the code needed on the python side is in his IL PR here: https://github.com/replicahq/model/pull/3117

If it's better to split that out into a separate PR, I'm happy to do that

**Review by kaelgreco — APPROVED** [review]

**Pull request merged**